code/M3D/M3DExporter.cpp

```diff
@@ -0,0 +1,394 @@
+    // Worker function for exporting a scene to binary M3D.
+    // Prototyped and registered in Exporter.cpp
+    void ExportSceneM3D (
+        const char* pFile,
+        IOSystem* pIOSystem,
+        const aiScene* pScene,
+        const ExportProperties* pProperties
+    ){
+        // initialize the exporter
+        M3DExporter exporter(pScene, pProperties);
+
+        // perform binary export
+        exporter.doExport(pFile, pIOSystem, false);
+    }
+
+    // ---------------------------------------------------------------------
+    // Worker function for exporting a scene to ASCII A3D.
+    // Prototyped and registered in Exporter.cpp
+    void ExportSceneA3D (
+        const char* pFile,
+        IOSystem* pIOSystem,
+        const aiScene* pScene,
+        const ExportProperties* pProperties
+
+    ){
+        // initialize the exporter
+        M3DExporter exporter(pScene, pProperties);
+
+        // perform ascii export
+        exporter.doExport(pFile, pIOSystem, true);
+    }
+
+} // end of namespace Assimp
+
+// ------------------------------------------------------------------------------------------------
+M3DExporter::M3DExporter ( const aiScene* pScene, const ExportProperties* pProperties )
+: mScene(pScene)
+, mProperties(pProperties)
+, outfile()
+, m3d(nullptr) { }
+
+// ------------------------------------------------------------------------------------------------
+void M3DExporter::doExport (
+    const char* pFile,
+    IOSystem* pIOSystem,
+    bool toAscii
+){
+    // TODO: convert mProperties into M3D_EXP_* flags
+    (void)mProperties;
+
+    // open the indicated file for writing (in binary / ASCII mode)
+    outfile.reset(pIOSystem->Open(pFile, toAscii ? "wt" : "wb"));
+    if (!outfile) {
+        throw DeadlyExportError( "could not open output .m3d file: " + std::string(pFile) );
+    }
+
+    // use malloc() here because m3d_free() will call free()
+    m3d = (m3d_t*)calloc(1, sizeof(m3d_t));
+    if(!m3d) {
+        throw DeadlyExportError( "memory allocation error" );
+    }
+    m3d->name = _m3d_safestr((char*)&mScene->mRootNode->mName.data, 2);
+
+    // Create a model from assimp structures
+    aiMatrix4x4 m;
+    NodeWalk(mScene->mRootNode, m);
+
+    // serialize the structures
+    unsigned int size;
+    unsigned char *output = m3d_save(m3d, M3D_EXP_FLOAT,
+        M3D_EXP_EXTRA | (toAscii ? M3D_EXP_ASCII : 0), &size);
+    m3d_free(m3d);
+    if(!output || size < 8) {
+        throw DeadlyExportError( "unable to serialize into Model 3D" );
+    }
+
+    // Write out serialized model
+    outfile->Write(output, size, 1);
+
+    // explicitly release file pointer,
+    // so we don't have to rely on class destruction.
+    outfile.reset();
+}
+
+// ------------------------------------------------------------------------------------------------
+// recursive node walker
+void M3DExporter::NodeWalk(const aiNode* pNode, aiMatrix4x4 m)
+{
+    unsigned int i, j, k, l, n, idx;
```

Comment:
I've minimized already, and the code passes the Quality Review. If they are initialized there and the Quality Review said it's fine that's because it is fine. Just saying.